vfs_handlers/s3_handlers.cc

```diff
@@ -425,306 +549,266 @@ makeStreamingDownload(const std::string & bucket,
     return makeStreamingDownload("s3://" + bucket + "/" + object);
 }
 
-struct StreamingUploadSource {
 
-    StreamingUploadSource(const std::string & urlStr,
-                          const OnUriHandlerException & excCallback,
-                          const S3Api::ObjectMetadata & metadata)
-    {
-        impl.reset(new Impl());
-        impl->owner = getS3ApiForUri(urlStr);
-        std::tie(impl->bucket, impl->object) = S3Api::parseUri(urlStr);
-        impl->metadata = metadata;
-        impl->onException = excCallback;
-        impl->chunkSize = 8 * 1024 * 1024;  // start with 8MB and ramp up
+/****************************************************************************/
+/* S3 UPLOADER                                                              */
+/****************************************************************************/
 
-        impl->start();
-    }
+inline void touchByte(const char * c)
+{
+    __asm__(" # [in]":: [in] "r" (*c):);
+}
 
-    typedef char char_type;
-    struct category
-        : public boost::iostreams::output,
-          public boost::iostreams::device_tag,
-          public boost::iostreams::closable_tag
-    {
-    };
+inline void touch(const char * start, size_t size)
+{
+    const char * current = start - (intptr_t) start % 4096;
+    if (current < start) {
+        current += 4096;
+    }
+    const char * end = start + size;
+    for (; current < end; current += 4096) {
+        touchByte(current);
+    }
+}
 
-    struct Impl {
-        Impl()
-            : offset(0), chunkIndex(0), shutdown(false),
-              chunks(16)
-        {
-        }
 
-        ~Impl()
-        {
-            //cerr << "destroying streaming upload at " << object << endl;
-            stop();
+struct S3Uploader {
+    S3Uploader(const S3Api * api,
+               const string & bucket,
+               const string & resource, // starts with "/", unescaped (buggy)
+               const OnUriHandlerException & excCallback,
+               const S3Api::ObjectMetadata & objectMetadata)
+        : api(api),
+          bucket(bucket), resource(resource),
+          metadata(objectMetadata),
+          onException(excCallback),
+          closed(false),
+          chunkSize(8 * 1024 * 1024), // start with 8MB and ramp up
+          currentRq(0),
+          activeRqs(0)
+    {
+        /* Maximum chunk size is what we can do in 3 seconds, up to 1% of
+           system memory. */
+#if 0
```

Comment:
I cannot remember, possibly a left over from a test. However, I wonder to what extend this is useful to use larger chunk sizes. Considering the limit of more or less 450mb/s we reach under specific conditions, for a hypothetical max chunk size of 50mbyte, that would represent 9 reqs/second and the overhead in that case is negligible.